### ISSUE TITLE
一个很奇怪的现象

### ISSUE THREAD

**Issue by Shadowhack0419** [shadowsocks/shadowsocks-libev#1600]
Please answer these questions before submitting your issue. Thanks!

(Please mention that if the issue you filed is solved, you may wish to close it by yourself. Thanks again.)

(PS, you can remove 3 lines above, including this one, before post your issue.)

### What version of shadowsocks-libev are you using?
3.0.8

### What operating system are you using?
Ubuntu16.04

### What did you do?
正常启动

### What did you expect to see?
速度在所有客户端表现相近

### What did you see instead?
使用libev部署服务器端在本地使用ss-local速度非常慢，但是使用安卓客户端速度很快，但是使用ssr部署服务端，安卓客户端会有限速现象，但是ss-local速度正常，会不会是实现差异导致了表现不同？

### What is your config in detail (with all sensitive info masked)?
正常配置使用AEAD加密

**Comment by madeye** [comment]
1. Try add `-d 8.8.8.8` to your server's command line.
2. Post the screenshots of `fast.com` with different clients and servers.

**Comment by Shadowhack0419** [comment]
@madeye -d参数之前已经加入，测试结果ss-local是手机端一半，在相同wifi条件下，实际情况看YouTube速度更小，图片这个客户端不太好发送，我以邮件形式发给你吧。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by baggiogogo** [comment]
手机浏览器是否安卓chrome，是的话检查下是否wifi走udp443直连了，我也最近发现这个问题，google上不去，后来发现走udp了，是的话转发443就可以了。

**Comment by Shadowhack0419** [comment]
@baggiogogo Google 上的去而且是手机端快，Linux的ss-local慢，并不是走udp443。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by madeye** [comment]
I cannot reproduce your.issue locally. Actually shadowsocks-android also uses ss-local.

My suggestion is checking your configs and system.carefully, make sure you are doing an apple-to-apple comparison.

**Comment by Shadowhack0419** [comment]
@madeye 测试地点就在我家WiFi旁边，测试方法为断开所以设备就，只连接测试设备测试，其他包括DNS等都相同，启动参数也相同。还有请问您的本地测试环境是在中国境内么？您的运营商是什么？我的是上海电信。:-)

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by madeye** [comment]
Please provide more details if you want to get your problem reproduced, including all the detailed configs for your PC and phone.

**Comment by Shadowhack0419** [comment]
@madeye 我的手机是MI5，电脑是DELL inspiron15 5000。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by terrytw** [comment]
你最好说具体一点
手机用的什么版本，设置截图贴一下，IP打码就行了
电脑用的什么版本，同样截图贴一下
电脑客户端应该有日志的，日志也贴一下吧

想要别人帮忙解决问题自己也得付出点努力提供足够的信息啊

**Comment by Shadowhack0419** [comment]
@terrytw 都是最新版，之前说了，手机配置：
路由：全局
远程dns:8.8.8.8
ipv6，dns转发全部关闭，未使用任何插件。
电脑配置相同，超时300其他没有任何变化，同样未使用插件
Linux PC日志：
2017-07-29 13:00:37 INFO: setting NOFILE to 512000
 2017-07-29 13:00:37 INFO: initializing ciphers... aes-128-gcm
 2017-07-29 13:00:37 INFO: tcp port reuse enabled
 2017-07-29 13:00:37 INFO: udprelay enabled
 2017-07-29 13:00:37 INFO: udp port reuse enabled
 2017-07-29 13:00:37 INFO: listening at 0.0.0.0:2333
 2017-07-29 13:00:52 INFO: connect to mtalk.google.com:5228
 2017-07-29 13:00:58 INFO: connect to www.google.com:443
 2017-07-29 13:00:58 INFO: connect to www.google.com.hk:443
 2017-07-29 13:00:58 INFO: connect to www.google.com.hk:80
 2017-07-29 13:00:58 INFO: connect to www.google.com.hk:80
 2017-07-29 13:00:58 INFO: connect to www.google.com.hk:443
 2017-07-29 13:00:59 INFO: connect to www.googleapis.com:443
 2017-07-29 13:00:59 INFO: connect to www.googleapis.com:443
 2017-07-29 13:00:59 INFO: connect to accounts.google.com:443
 2017-07-29 13:01:00 INFO: connect to clients2.google.com:443
 2017-07-29 13:01:01 INFO: connect to accounts.google.com:443
 2017-07-29 13:01:01 INFO: connect to clients4.google.com:443
 2017-07-29 13:01:01 INFO: connect to clients4.google.com:443
 2017-07-29 13:01:03 INFO: connect to accounts.google.com:443
2017-07-29 13:01:14 INFO: connect to notifications.google.com:443
 2017-07-29 13:01:14 INFO: connect to play.google.com:443
 2017-07-29 13:01:17 INFO: connect to play.google.com:443
 2017-07-29 13:01:21 INFO: connect to www.youtube.com:443
 2017-07-29 13:01:25 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:25 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:25 INFO: connect to s0.2mdn.net:443
 2017-07-29 13:01:25 INFO: connect to securepubads.g.doubleclick.net:443
 2017-07-29 13:01:25 INFO: connect to tpc.googlesyndication.com:443
 2017-07-29 13:01:25 INFO: connect to pubads.g.doubleclick.net:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:25 INFO: connect to i.ytimg.com:443
 2017-07-29 13:01:26 INFO: connect to yt3.ggpht.com:443
 2017-07-29 13:01:26 INFO: connect to yt3.ggpht.com:443
 2017-07-29 13:01:27 INFO: connect to s0.2mdn.net:443
 2017-07-29 13:01:27 INFO: connect to pagead2.googlesyndication.com:443
 2017-07-29 13:01:27 INFO: connect to googleads4.g.doubleclick.net:443
doubleclick.net:443
 2017-07-29 13:01:29 INFO: connect to www.gstatic.com:443
 2017-07-29 13:01:29 INFO: connect to t.mookie1.com:443
 2017-07-29 13:01:31 INFO: connect to tlg.mookie1.com:443
 2017-07-29 13:01:32 INFO: connect to ad.doubleclick.net:443
 2017-07-29 13:01:33 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:33 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:33 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:33 INFO: connect to fonts.gstatic.com:443
 2017-07-29 13:01:33 INFO: connect to secure.insightexpressai.com:443
 2017-07-29 13:01:33 INFO: connect to secure.insightexpressai.com:443
 2017-07-29 13:01:33 INFO: connect to static.doubleclick.net:443
 2017-07-29 13:01:35 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:35 INFO: connect to s.youtube.com:443
 2017-07-29 13:01:35 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:35 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:35 INFO: connect to r4---sn-ab5l6nzr.googlevideo.com:443
 2017-07-29 13:01:35 INFO: connect to r4---sn-ab5l6nzr.googlevideo.com:443
 2017-07-29 13:01:37 INFO: connect to r9---sn-ab5l6n7e.googlevideo.com:443
 2017-07-29 13:01:37 INFO: connect to r9---sn-ab5l6n7e.googlevideo.com:443
 2017-07-29 13:01:37 ERROR: server_recv_cb_recv: Connection reset by peer
 2017-07-29 13:01:37 INFO: connect to r9---sn-ab5l6n7e.googlevideo.com:443

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Shadowhack0419** [comment]
我通过实时监控日志发现libev部署服务端用ss-local连接看YouTube视频，一直卡住的时候它一直在切换连接节点，每次连接十几秒后又更换，没有一点速度。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by madeye** [comment]
I still cannot find anything useful from the information you provided.

What you need to provide includes:

1. The *detailed* JSON config and command line on your server.
2. The *detailed* JSON config and command line on your client.
3. The *detailed* steps to reproduce your "stucked" connection.

**Comment by Shadowhack0419** [comment]
@madeye server config.json:

{
    "server":["[::0]", "0.0.0.0"],
    "server_port":9071,
    "password":"xxxxx",
    "method":"aes-128-gcm",
    "mode": "tcp_and_udp",
    "ipv6_first": false,
    "user": "nobody",
    "reuse_port": true,
    "nofile": 512000,
    "timeout":300,
    "fast_open": false
}

Client:
{
    "server":"xxxx",
    "server_port":9071,
    "local_address":"0.0.0.0",
    "local_port":2333,
    "password":"xxxxx",
    "timeout":300,
    "method":"aes-128-gcm",
    "fast_open": false
}

Reproduce everytime I watch the video on Youtube with my Ubuntu PC.

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by madeye** [comment]
1. Open `chrome://flags/#enable-quic` in your Chrome browser, and post the settings here.
2. Try Firefox on your PC.
3. Remove all the `--reuse-port` and `-u` options from your config and command line.

**Comment by Shadowhack0419** [comment]
@madeye 1.打开后里面的实验性功能全部是默认，由于手机问题，无法发出图片，抱歉。
2.尝试移除所以参数，使用firefox浏览器也是同样结果。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Shadowhack0419** [comment]
@madeye 经过测试，如果安卓使用NAT模式连接也会出现Linux的ss-local卡住现象，所以请问会不会是安卓的VPN模式造成了GFW所没有发现的漏洞?

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by madeye** [comment]
1. Change the server port from 9071 to 443.
2. Change 'chrome://flags/#enable-quic' to disabled.

Then try again.

**Comment by madeye** [comment]
Change your server config to:

```
{
"server":"0.0.0.0",
"server_port":443,
"password":"xxxxx",
"method":"aes-128-gcm",
"nofile": 512000,
"timeout":300
}
```

Change your server command line to:

```
ss-server -c /etc/ss.json -f /tmp/pid
```

Then try:

1. Disable NAT mode in your Android app and open Youtube app to play a video.
2. Enable NAT mode in your Android app and open Youtube app to play the same video.

Post the screenshot of "nerd stats" here.

**Comment by madeye** [comment]
Since
1. NAT mode doesn't support UDP forwarding.
2. Chrome doesn't support SOCKS5 UDP relay.

Then,
1. With NAT mode disabled, Youtube app would use QUIC (a UDP protocol) to transfer videos, which is expected to be faster.
2. With NAT mode enabled, Youtube app can only transfer data in HTTPS, which is expected to be slower than QUIC.
3. On your PC, Chrome can only transfer data in HTTPS via ss-local's SOCKS5 proxy, which is expected to have the same performance with 2).

So, in summary, your observation is expected and has nothing to do with shadowsocks.

To improve the performance, you may try:

1. kcptun
2. BBR
3. BBR + simple-obfs.

**Comment by Shadowhack0419** [comment]
@madeye 那么请问有什么方式可是强制使chorme强制使用udp流量？

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Shadowhack0419** [comment]
问题解决，是内核参数调整的问题导致tcp连接缓慢。

_Sent from my Xiaomi MI 5 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by GreenSkinMonster** [comment]
@Shadowhack0419 方便说一下是哪个内核参数吗？谢谢。
